File: packages/sources/sdk-py/src/mplp/learning/collector.py
from typing import List, Optional
from ..observability.types import MplpEvent
from .types import LearningSample
from .hooks import collect_learning_sample
from .validator import validate_learning_sample

class LearningCollector:
    def __init__(self, enabled: bool = False, max_buffer: int = 1000):
        self.enabled = enabled
        self.max_buffer = max_buffer
        self.buffer: List[LearningSample] = []

    def set_enabled(self, enabled: bool):
        self.enabled = enabled

    def on_event(self, event: MplpEvent):
        if not self.enabled:
            return

        sample = collect_learning_sample(event)
        if sample:
            # Validate immediately (Fail-Fast / Drop Invalid)
            validation = validate_learning_sample(sample)
            if validation.valid:
                if len(self.buffer) < self.max_buffer:
                    self.buffer.append(sample)
                else:
                    print("[LearningCollector] Buffer full, dropping sample.")
            else:
                print(f"[LearningCollector] Generated invalid sample: {validation.errors}")

    def flush(self) -> List[LearningSample]:
        samples = list(self.buffer)
        self.buffer = []
        return samples

    def clear(self):
        self.buffer = []

File: packages/sources/sdk-ts/src/learning/collector.ts
import { MplpEvent } from '../observability/types';
import { LearningSample } from './types';
import { collectLearningSample } from './hooks';
import { validateLearningSample } from './validator';

export interface LearningCollectorConfig {
    enabled: boolean;
    maxBuffer?: number;
}

export class LearningCollector {
    private enabled: boolean;
    private maxBuffer: number;
    private buffer: LearningSample[] = [];

    constructor(config: LearningCollectorConfig) {
        this.enabled = config.enabled;
        this.maxBuffer = config.maxBuffer || 1000;
    }

    public setEnabled(enabled: boolean): void {
        this.enabled = enabled;
    }

    public onEvent(event: MplpEvent): void {
        if (!this.enabled) return;

        const sample = collectLearningSample(event);
        if (sample) {
            // Validate immediately (Fail-Fast / Drop Invalid)
            // Per Route B requirements: "validator PASS"
            const validation = validateLearningSample(sample);
            if (validation.valid) {
                if (this.buffer.length < this.maxBuffer) {
                    this.buffer.push(sample);
                } else {
                    // Drop policy: simple drop for now, or rotate?
                    // "drop_policy" was mentioned in user request.
                    // We'll just drop new ones to prevent OOM in this minimal implementation.
                    console.warn("[LearningCollector] Buffer full, dropping sample.");
                }
            } else {
                console.warn(`[LearningCollector] Generated invalid sample: ${validation.errors.map(e => e.message).join(', ')}`);
            }
        }
    }

    public flush(): LearningSample[] {
        const samples = [...this.buffer];
        this.buffer = [];
        return samples;
    }

    public clear(): void {
        this.buffer = [];
    }
}
